chrpath.1
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.\"
.TH CHRPATH 1 "May 4, 2002" "chrpath" "change rpath/runpath in binaries"
.SH NAME
chrpath \- change the rpath or runpath in binaries
.SH SYNOPSIS
\fBchrpath\fP
[ \fB-v\fP | \fB--version\fP ]
[ \fB-d\fP | \fB--delete\fP ]
[ \fB-r\fP \fI<path>\fP |\  \fB--replace\fP \fI<path>\fP ]
[ \fB-c\fP | \fB--convert\fP ]
[ \fB-l\fP | \fB--list\fP ]
[ \fB-h\fP | \fB--help\fP ]
\fI<program>\fP [ \fI<program>\fP ... ]
.SH DESCRIPTION
\fBchrpath\fP changes, lists or removes the rpath or runpath setting in a
binary.  The rpath, or runpath if it is present, is where the runtime linker
should look for the libraries needed for a program.
.SH OPTIONS
.TP
.BR -v \ |\  --version
Display program version number
.TP
.BR -d \ |\  --delete
Delete current rpath or runpath setting
.TP
.BR -c \ |\  --convert
Convert the rpath setting into a runpath setting
.TP
.BI -r\ \fI<path>\fP\ |\ --replace\ \fI<path>\fP
Replace current rpath or runpath setting with the path given.  The new
path must be shorter or the same length as the current path.
.TP
.BR -k \ |\  --keepgoing
Do not fail on first error, but process all arguments before returning
the error.
.TP
.BR -l \ |\  --list
List the current rpath or runpath (default)
.TP
.BR -h \ |\  --help
Show usage information.
.SH EXIT STATUS
.TP
.BR 0
If all operations were successful
.TP
.BR > 0
if one of the operations failed.  A failing operation terminates
the program unless -k is specified.

.SH BUGS
This program cannot create an RPATH tag if the ELF does not have one,
and it can only replace an RPATH with one of equal or shorter
length. (Moving ELF sections following a lengthened string table would
be difficult and error-prone at best, and is sometimes outright
impossible due to issues like limited ranges in jump instructions.)

.SH AUTHOR
The chrpath program was written by Petter Reinholdtsen
<pere@hungry.com>, based on works by Geoffrey Keating
<geoffk@ozemail.com.au> and Peeter Joot <peeterjoot@protonmail.com>.

This manual page was originally written by
Tollef Fog Heen <tfheen@debian.org>, for the Debian GNU/Linux
system (but may be used by others).


NEWS
New in 0.17 released 2014-XX-XX:
 * Add Solaris support (use <sys/byteorder.h> instead of <byteswap.h>).
   Patch from Rainer Orth.

New in 0.16 released 2014-01-14:
 * Fixed all minor bugs discovered by Coverity.
 * Updated config.sub and config.guess from the GNU project.
 * Mention new project mailing list in the documentation.

New in 0.15 released 2013-11-24:
 * Updated config.sub and config.guess from the GNU project to work with
   newer architectures.  Thanks to isha vishnoi for the heads up.
 * Updated README with current URLs.
 * Added byteswap fix found in Ubuntu, credited Jeremy Kerr and
   Matthias Klose.
 * Added missing help for -k|--keepgoing option, using patch by Petr
   Machata found in Fedora.
 * Rewrite removal of RPATH/RUNPATH to make sure the entry in .dynamic
   is a NULL terminated string.  Based on patch found in Fedora
   credited Axel Thimm and Christian Krause.

New in 0.14 released 2011-09-27:
 * Updated debian build rules from the current Debian package.
 * Applied patch from debian to support Hurd and k*BSD.
 * Moved project to <URL: https://alioth.debian.org/projects/chrpath/ >
 * Updated autoconf/automake-related files to latest versions.
 * Document rpath size limitation better in the manual page.
   Patch from Eric Raymond.
 * Fix typo in handling of long arguments.  Patch from Eric Raymond.
 * Now works even when the ELF file has endianness/word length 
   different from the host machime.  Patch from Javier Serrano Polo,
   adjusted to compile with -ansi and -pedantic.

New in 0.13 released 2004-09-19:
 * Change exit code handling.  Return an error if the operation fail.
 * New argument -k (--keepgoing) to not stop on first error.
 * Improve test code.
 * Fix typo in manual page.

New in 0.12 released 2003-12-12:
 * Use p_filesz instead p_memsz when checking if the elf
   header exist.  This should get it working on Solaris.
 * Try to get it compiling on Mac OS X / Darwin.

New in 0.11 released 2003-09-15:
 * Make chrpath work better on 64 bit arches.
 * Try to make the test code more portable (choosing
   -rpath/-R depending on arch).

New in 0.10 released 2002-11-30:
 * Avoid segfaulting when derefering a NULL pointer.
 * Added simple self test code.
 * Converted to use newer autoconf (2.53).

New in 0.9 released 2002-10-28:
 * Bugfix in option parsing.  Now only --replace takes a parameter.
 * Port to Irix 6.5.
 * Avoid compile warning on platforms without getopt_long().

New in 0.8 released 2002-09-17:
 * Handle platforms without runpath support (ie missing DT_RUNPATH).
 * Fix compile problem on Solaris 2.6.

New in 0.7 released 2002-08-01:
 * Fixed small memleak.
 * Fixed crash bug on Solaris, when reading unknown/broken formats.
 * Fix typos in manpage (s/--remove/--replace/ and some s/ rpath/ chrpath/).
 * Added runpath-support.

New in 0.6 released 2002-05-05:
 * Added manual page, thanks to Tollef Fog Heen.
 * Fix file descriptor leak when elf_open fails.
 * Update URLs and contact information in documentation.
 * Added debian/ package scripts.

New in 0.5 released 2001-11-03:
 * Compile fixes for platforms without getopt_long().
 * Handle more then one file on one run.

New in 0.4 released 2001-10-25:

 * Fixes for SPARC/Solaris.
 * Corrected link to elfhack.

New in 0.3 released 2001-10-16:

 * Fix typo in ELF header reading.
 * Get rid of some compiler warnings.

New in 0.2 released 2001-10-12:

 * Fixed some typos in option parsing.
 * Added link to elfhack.

New in 0.1 released 2001-03-01:

 * First release
 * Support deleting and modifying the rpath information


chrpath.c
/*
<URL:http://gcc.gnu.org/ml/gcc/1999-04n/msg01105.html>

Re: changing embedded RPATH in existing executables.

  To: geoffk@ozemail.com.au
  Subject: Re: changing embedded RPATH in existing executables.
  From: <peeter_joot@VNET.IBM.COM> (peeter joot)
  Date: Fri, 30 Apr 1999 16:14:44 -0400 (EDT)
  Cc: peeterj@ca.ibm.com, egcs@cygnus.com, libc-hacker@cygnus.com, linux-gcc@vger.rutgers.edu
  Reply-To: <peeter_joot@VNET.IBM.COM>

> _Changing_ is a little tricky, but the attached program strips rpaths
> from executables (I find it essential for debugging the binutils).
> It's endian-dependent, if you want this for x86 you can just change
> the occurrences of 'MSB' to 'LSB' and compile (I should really fix
> that).

Hi Geoff,

With your program as a guide (and some peeks into libbfd, elf.h, a bit
of the glibc dynamic loader code, objdump, and a hex-editor) I was able to
figure out enough to find and change the rpath string.  That was fun!

This program assumes (unlike your original program) that there is only
one DT_RPATH tag in the dynamic section as even with multiple '-Wl,-rpath,'
commands in the link this seems to occur (they all get concatonated into
a : separated path).

Thanks for your help.  If you want to use this on non-x86 you have to change
the occurances of LSB back to MSB:)

Peeter
--
*/

#ifdef HAVE_CONFIG_H
#  include "config.h"
#endif

#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <elf.h>
#if defined(HAVE_LINK_H)
#  include <link.h>
#endif /* HAVE_LINK_H */
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include "protos.h"

/**
 * Reads an ELF file, and reads or alters the RPATH setting.
 *
 * TODO:
 *  modify to add RPATH setting if none exists.
 */


int
chrpath(const char *filename, const char *newpath, int convert)
{
  int fd;
  Elf_Ehdr ehdr;
  int i;
  Elf_Phdr phdr;
  Elf_Shdr shdr;
  void *dyns;
  int rpathoff;
  char * strtab;
  char * rpath;
  unsigned int rpathlen;
  int oflags;
  int rpath_dyns_index;

  if (NULL == newpath && 0 == convert)
     oflags = O_RDONLY;
  else
     oflags = O_RDWR;

  fd = elf_open(filename, oflags, &ehdr);
  if (fd == -1)
  {
    perror ("elf_open");
    return 1;
  }

   if (0 != elf_find_dynamic_section(fd, &ehdr, &phdr))
   {
     perror("found no dynamic section");
     elf_close(fd);
     return 1;
   }

  dyns = malloc(PHDR(p_filesz));
  if (dyns == NULL)
    {
      perror ("allocating memory for dynamic section");
      elf_close(fd);
      return 1;
    }
  memset(dyns, 0, PHDR(p_filesz));
  if (lseek(fd, PHDR(p_offset), SEEK_SET) == -1
      || read(fd, dyns, PHDR(p_filesz)) != (ssize_t)PHDR(p_filesz))
    {
      perror ("reading dynamic section");
      free(dyns);
      elf_close(fd);
      return 1;
    }

  rpathoff = -1;
  for ( rpath_dyns_index = 0; DYNSS(rpath_dyns_index, d_tag) != DT_NULL;
        ++rpath_dyns_index )
    {
      if ( elf_dynpath_tag(DYNSS(rpath_dyns_index, d_tag)) )
      {
         rpathoff = DYNSU(rpath_dyns_index, d_un.d_ptr);
         break;
      }
    }
  if (rpathoff == -1)
    {
      printf("%s: no rpath or runpath tag found.\n", filename);
      free(dyns);
      elf_close(fd);
      return 2;
    }

  if (lseek(fd, EHDRWU(e_shoff), SEEK_SET) == -1)
  {
    perror ("positioning for sections");
    free(dyns);
    elf_close(fd);
    return 1;
  }

  for (i = 0; i < EHDRHU(e_shnum); i++)
  {
    const size_t sz_shdr = is_e32() ? sizeof(Elf32_Shdr) : sizeof(Elf64_Shdr);
    if (read(fd, &shdr, sz_shdr) != (ssize_t)sz_shdr)
    {
      perror ("reading section header");
      free(dyns);
      elf_close(fd);
      return 1;
    }
    if (SHDR_W(sh_type) == SHT_STRTAB)
      break;
  }
  if (i == EHDRHU(e_shnum))
    {
      fprintf (stderr, "No string table found.\n");
      free(dyns);
      elf_close(fd);
      return 2;
    }
  /* +1 for forced trailing null */
  strtab = (char *)malloc(SHDR_O(sh_size)+1);
  if (strtab == NULL)
    {
      perror ("allocating memory for string table");
      free(dyns);
      elf_close(fd);
      return 1;
    }
  memset(strtab, 0, SHDR_O(sh_size)+1);

  if (lseek(fd, SHDR_O(sh_offset), SEEK_SET) == -1)
  {
    perror ("positioning for string table");
    free(strtab);
    free(dyns);
    elf_close(fd);
    return 1;
  }
  if (read(fd, strtab, SHDR_O(sh_size)) != (ssize_t)SHDR_O(sh_size))
  {
    perror ("reading string table");
    free(strtab);
    free(dyns);
    elf_close(fd);
    return 1;
  }

  if ((int)SHDR_O(sh_size) < rpathoff)
  {
    fprintf(stderr, "%s string offset not contained in string table",
            elf_tagname(DYNSS(rpath_dyns_index, d_tag)));
    free(strtab);
    free(dyns);
    elf_close(fd);
    return 5;
  }
  rpath = strtab+rpathoff;

#if defined(DT_RUNPATH)
  if (convert && DYNSS(rpath_dyns_index, d_tag) == DT_RPATH)
  {
    if (is_e32())
      ((Elf32_Dyn *)dyns)[rpath_dyns_index].d_tag = swap_bytes() ?
        bswap_32(DT_RUNPATH) : DT_RUNPATH;
    else
      ((Elf64_Dyn *)dyns)[rpath_dyns_index].d_tag = swap_bytes() ?
        bswap_64(DT_RUNPATH) : DT_RUNPATH;
    if (lseek(fd, PHDR(p_offset), SEEK_SET) == -1
        || write(fd, dyns, PHDR(p_filesz)) != (int)PHDR(p_filesz))
    {
      perror ("converting RPATH to RUNPATH");
      free(strtab);
      free(dyns);
      elf_close(fd);
      return 1;
    }
    printf("%s: RPATH converted to RUNPATH\n", filename);
  }
#endif /* DT_RUNPATH */

  printf("%s: %s=%s\n", filename, elf_tagname(DYNSS(rpath_dyns_index, d_tag)),
         rpath);

  if (NULL == newpath)
  {
    free(dyns);
    free(strtab);
    elf_close(fd);
    return 0;
  }

  rpathlen = strlen(rpath);

  /*
   * Calculate the maximum rpath length (will be equal to rpathlen unless
   * we have previously truncated it).
   */
  for ( i = rpathoff + rpathlen ; (i < (int)SHDR_O(sh_size)
                                   && strtab[i] == '\0') ; i++ )
    ;
  i--;

  if (i > (int)(rpathoff + rpathlen))
     rpathlen = i - rpathoff;

  if (strlen(newpath) > rpathlen)
  {
    fprintf(stderr, "new rpath '%s' too large; maximum length %i\n",
            newpath, rpathlen);
    free(dyns);
    free(strtab);
    elf_close(fd);
    return 7;
  }

  memset(rpath, 0, rpathlen);
  strcpy(rpath, newpath);

  if (lseek(fd, SHDR_O(sh_offset)+rpathoff, SEEK_SET) == -1)
  {
    perror ("positioning for RPATH");
    free(dyns);
    free(strtab);
    elf_close(fd);
    return 1;
  }
  if (write(fd, rpath, rpathlen) != (int)rpathlen)
  {
    perror ("writing RPATH");
    free(dyns);
    free(strtab);
    elf_close(fd);
    return 1;
  }
  printf("%s: new %s: %s\n", filename,
         elf_tagname(DYNSS(rpath_dyns_index, d_tag)), rpath);

  elf_close(fd);

  free(dyns);
  dyns = NULL;

  free(strtab);

  return 0;
}
